instrumentation/undertow-1.4/src/main/java/io/opentelemetry/javaagent/instrumentation/hypertrace/undertow/v1_4/StreamSourceChannelInstrumentation.java

```diff
@@ -0,0 +1,89 @@
+
+import static net.bytebuddy.matcher.ElementMatchers.isPublic;
+import static net.bytebuddy.matcher.ElementMatchers.named;
+import static net.bytebuddy.matcher.ElementMatchers.returns;
+import static net.bytebuddy.matcher.ElementMatchers.takesArgument;
+import static net.bytebuddy.matcher.ElementMatchers.takesArguments;
+
+import io.opentelemetry.javaagent.instrumentation.api.ContextStore;
+import io.opentelemetry.javaagent.instrumentation.api.InstrumentationContext;
+import io.opentelemetry.javaagent.instrumentation.hypertrace.undertow.v1_4.utils.Utils;
+import io.opentelemetry.javaagent.tooling.TypeInstrumentation;
+import io.opentelemetry.javaagent.tooling.bytebuddy.matcher.AgentElementMatchers;
+import java.nio.ByteBuffer;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import net.bytebuddy.asm.Advice;
+import net.bytebuddy.description.method.MethodDescription;
+import net.bytebuddy.description.type.TypeDescription;
+import net.bytebuddy.matcher.ElementMatcher;
+import net.bytebuddy.matcher.ElementMatcher.Junction;
+import org.hypertrace.agent.core.instrumentation.HypertraceCallDepthThreadLocalMap;
+import org.hypertrace.agent.core.instrumentation.SpanAndObjectPair;
+import org.xnio.channels.StreamSourceChannel;
+
+public final class StreamSourceChannelInstrumentation implements TypeInstrumentation {
+
+  @Override
+  public ElementMatcher<TypeDescription> typeMatcher() {
+    return AgentElementMatchers.safeHasSuperType(named("org.xnio.channels.StreamSourceChannel"));
+  }
+
+  @Override
+  public Map<? extends ElementMatcher<? super MethodDescription>, String> transformers() {
+    final Map<Junction<MethodDescription>, String> matchers = new HashMap<>();
+    matchers.put(
+        named("read")
+            .and(takesArguments(1))
+            .and(takesArgument(0, ByteBuffer.class))
+            .and(returns(int.class))
+            .and(isPublic()),
+        StreamSourceChannelInstrumentation.class.getName() + "$read_advice");
+    return Collections.unmodifiableMap(matchers);
+  }
+
+  /** TODO */
+  static final class read_advice {
+
+    @Advice.OnMethodEnter
+    public static int trackCallDepth() {
+      return HypertraceCallDepthThreadLocalMap.incrementCallDepth(StreamSourceChannel.class);
+    }
+
+    @Advice.OnMethodExit(suppress = Throwable.class)
+    public static void enter(
+        @Advice.Enter int callDepth,
+        @Advice.Return final int numBytesRead,
+        @Advice.This final StreamSourceChannel streamSourceChannel,
+        @Advice.Argument(0) final ByteBuffer byteBuffer) {
+      if (callDepth > 0) {
+        return;
+      }
+      HypertraceCallDepthThreadLocalMap.reset(StreamSourceChannel.class);
+      final ContextStore<StreamSourceChannel, SpanAndObjectPair> contextStore =
+          InstrumentationContext.get(StreamSourceChannel.class, SpanAndObjectPair.class);
+      final SpanAndObjectPair spanAndObjectPair = contextStore.get(streamSourceChannel);
```

Comment:
nit: remove `final` it does not help here and it just adds noise.

Reply:
I disagree here - I find that defaulting local variables to final discourages mutating references to objects and results in code that is easier to read and maintain.

Reply:
`final` is useful in large methods where there is a higher probability of making a mistake by accidentally reusing the variable. The `final` should be used to improve the readability of large code blocks or as a "documentation" of the design.
In this PR the final keyword is used for all local variables in very short methods (e.g. 5-10 LOC), which does not make sense to me.

The objective is to write as little code as possible because that is the easiest to read, maintain.... Look at Java std lib/JDK implementation or read the effective java book.

Reply:
I'm a big fan of effective java! I don't recall an item about final and/or non-final local variables, but I personally find that flipping the switch in the IDE and defaulting to final local variables makes the code easier to maintain and build upon over time. The cost of making variables final is low (because it can be done automatically) and IMO it helps make the code a little more readable because its clear at variable declaration time when a variable is reassigned, rather than relying on the reader to figure out whether a variable is effectively final (because it's never reassigned) or if it's actually mutated. This scales well over time as a function organically grows in length over time and is modified by multiple authors, who may or may not have context on what the previous person wrote. I guess I just disagree that the least amount of code possible is the easiest to read and maintain. A good counterexample to that point, in my opinion, is how books like effective java encourage the use of inheritance over composition. In general, using composition rather than inheritance results in more code (as measured by LOC) but most folks agree that composition is a better default method of sharing code because it is easier to understand what is happening when read in the future. That's not to say that there isn't a place for inheritance (or non-final local variables), just that it's part of what influences what I tend to do by default. 


That being said, in the context of this PR it isn't terribly important, so if you feel strongly about this I'd be happy to table the discussion for another time and remove the `final` keyword from local variables in this PR. We can address it at a time when we can also add in automation to enforce whatever decision we come to.